### PR TITLE
[Feature](Nereids) Reorder join to eliminate cross join.

### PR DESCRIPTION
**Pull request by wangshuo128**
# Proposed changes

1. Try to eliminate cross join via finding join conditions in filters and changing the join orders.
  
For example:

 ``` sql
-- input:
SELECT * FROM t1, t2, t3 WHERE t1.id=t3.id AND t2.id=t3.id

-- output:
SELECT * FROM t1 JOIN t3 ON t1.id=t3.id JOIN t2 ON t2.id=t3.id
```
This feature is controlled by session variable `enable_nereids_reorder_to_eliminate_cross_join` with true by default.

2. Simplify usage of `Memo` and rewrite rule application.

Before this PR, if we want to apply a rewrite rule to a plan, the code is like the below:
```java
    Memo memo = new Memo();
        memo.initialize(root);

    PlannerContext plannerContext = new PlannerContext(memo, new ConnectContext());
    JobContext jobContext = new JobContext(plannerContext, new PhysicalProperties(), 0);
    RewriteTopDownJob rewriteTopDownJob = new RewriteTopDownJob(memo.getRoot(),
            ImmutableList.of(new AggregateDisassemble().build()), jobContext);
        plannerContext.pushJob(rewriteTopDownJob);
        plannerContext.getJobScheduler().executeJobPool(plannerContext);

    Plan after = memo.copyOut();
```
After this PR, we could use chain style calling:
```java
    new Memo(plan)
        .newPlannerContext(connectContext)
        .setDefaultJobContext()
        .topDownRewrite(new AggregateDisassemble())
        .getMemo()
        .copyOut();
```

3.  Rename the session variable `enable_nereids` to `enable_nereids_planner` to make it more meaningful.

## Problem Summary:

Describe the overview of changes.

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
5. Has document been added or modified: (Yes/No/No Need)
6. Does it need to update dependencies: (Yes/No)
7. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+ * SELECT * FROM t1 JOIN t3 ON t1.id=t3.id JOIN t2 ON t2.id=t3.id
+ * </pre>
+ * </p>
+ * TODO: This is tested by SSB queries currently, add more `unit` test for this rule
+ * when we have a plan building and comparing framework.
+ */
+public class ReorderJoin extends OneRewriteRuleFactory {
+    @Override
+    public Rule<Plan> build() {
+        return logicalFilter(subTree(LogicalJoin.class, LogicalFilter.class)).then(filter -> {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java:61]: do we need to add LogicalProject.class into list

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+            PlanCollector collector = new PlanCollector();
+            filter.accept(collector, null);
+            List<Plan> joinInputs = collector.joinInputs;
+            List<Expression> conjuncts = collector.conjuncts;
+
+            if (joinInputs.size() >= 3 && !conjuncts.isEmpty()) {
+                return reorderJoins(joinInputs, conjuncts);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java:68]: i think we chould add other reorder algorithm and use a config to decide witch algorithm should use here later. am i right?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+            } else {
+                return filter;
+            }
+        }).toRule(RuleType.REORDER_JOIN);
+    }
+
+    private Plan reorderJoins(List<Plan> joinInputs, List<Expression> conjuncts) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java:75]: please add comments to explain which join reorder algorithm is used here. And add algorithm name to function name

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+        HashSet<Slot> joinOutput = new HashSet<>();
+        joinOutput.addAll(left.getOutput());
+        joinOutput.addAll(right.getOutput());
+        return joinOutput;
+    }
+
+    private Map<Boolean, List<Expression>> splitConjuncts(List<Expression> conjuncts, Set<Slot> slots) {
+        return conjuncts.stream().collect(Collectors.partitioningBy(
+                // TODO: support non equal to conditions.
+                expr -> expr instanceof EqualTo && slots.containsAll(SlotExtractor.extractSlot(expr))));
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java:156]: maybe we need to check left child and right child slots on the different side of the equation

File: fe/fe-core/src/main/java/org/apache/doris/nereids/util/PlanRewriter.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+import org.apache.doris.nereids.memo.Memo;
+import org.apache.doris.nereids.rules.Rule;
+import org.apache.doris.nereids.rules.RuleFactory;
+import org.apache.doris.nereids.trees.plans.Plan;
+import org.apache.doris.qe.ConnectContext;
+
+/**
+ * Utility to copy plan into {@link Memo} and apply rewrite rules.
+ */
+public class PlanRewriter {
```
[the remark this answers is not in the document]
Comment (wangshuo128) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/util/PlanRewriter.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindRelation.java [hunk 6]
```diff
@@ -36,34 +37,34 @@
...
     public Rule<Plan> build() {
-        // fixme, just for example now
         return unboundRelation().thenApply(ctx -> {
             ConnectContext connectContext = ctx.plannerContext.getConnectContext();
             List<String> nameParts = ctx.root.getNameParts();
             switch (nameParts.size()) {
                 case 1: {
-                    List<String> qualifier = Lists.newArrayList(connectContext.getDatabase(), nameParts.get(0));
-                    Table table = getTable(qualifier, connectContext.getCatalog());
+                    List<String> qualified = ImmutableList.of(connectContext.getDatabase(), nameParts.get(0));
```
[the remark this answers is not in the document]
Comment (wangshuo128) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindRelation.java]: IIUC, `qualifier` is the prefix of `nameParts` except the last one.
Let's say we have a `nameParts` [catalog, db, table], its qualifier should be [catalog, db], right?
Please correct me if any goes wrong.
The `qualified` in the previous code was meant to fully qualified name.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java (hunk 2 above)
Comment (wangshuo128) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java:68]: Yes, any candidate config name? BTW,  I prefer a unified config name prefix in the new planner, such `enable_nereids_xxx`.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java (hunk 4 above)
Comment (wangshuo128) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java:156]: Actually, this doesn't make a big difference. 
This rule push predicates with slots from one input into the join. 
`PushPredicateThroughJoin` could push this kind of predicates through join, both in join conditions or filter predicates.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindRelation.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindRelation.java]: got it

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+ * SELECT * FROM t1 JOIN t3 ON t1.id=t3.id JOIN t2 ON t2.id=t3.id
+ * </pre>
+ * </p>
+ * TODO: This is tested by SSB queries currently, add more `unit` test for this rule
+ * when we have a plan building and comparing framework.
+ */
+public class ReorderJoin extends OneRewriteRuleFactory {
+    @Override
+    public Rule<Plan> build() {
+        return logicalFilter(subTree(LogicalJoin.class, LogicalFilter.class)).then(filter -> {
```
Comment (jackwener) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java:61]:
```suggestion
        return logicalFilter(subTree(LogicalJoin.class, LogicalFilter.class)).then(subTreeRoot -> {
```

Comment (924060929) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java:61]: the root is the filter, not the subTreeRoot

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+    }
+
+    private Set<Slot> getJoinOutput(Plan left, Plan right) {
+        HashSet<Slot> joinOutput = new HashSet<>();
+        joinOutput.addAll(left.getOutput());
+        joinOutput.addAll(right.getOutput());
+        return joinOutput;
+    }
+
+    private Map<Boolean, List<Expression>> splitConjuncts(List<Expression> conjuncts, Set<Slot> slots) {
```
Comment (jackwener) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java:167]: Can we move it to a new file `JoinUtils.java`?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java (hunk 2 above)
Comment (wangshuo128) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java]: Added a session variable `enable_nereids_reorder_to_eliminate_cross_join`.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java (hunk 8 above)
Comment (wangshuo128) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java:172]: I would refactor similar logic when we support a more comprehensive join and filter rewrite.

**Comment by 924060929** [comment]
LGTM

**Review by 924060929 — APPROVED** [review]
LGTM

**Pull request merged**